[Fix #1924] Use a cache to speed things up

jonas054 opened the pull request:
I've been working on this for a while, and feel ready to make a pull request. The performance of RuboCop has been discussed lately, so I guess this would be welcomed by some users.

Cache invalidation is generally a hard problem, but I think this solution should work (knock on wood). It uses the source code of the `rubocop` executable, a selection of the passed options, the inspected file, and its configuration as basis.

Add `--cache` option and `AlwaysRunWithCache` configuration parameter (`false` by default). Store a cache of inspection results at `/tmp/rubocop_cache/` (or whatever the temporary directory is). Clean up when it gets too big (governed by the configuration parameter `MaxFilesInCache`).

**Performace**

There's a slight performance penalty for storing cached results while inspecting files, around 2%, which is why I chose to only store in the cache if the option is given. The execution when there are cached results to reuse is typically between 10 and 30 times faster.

bbatsov commented:
Great feature! I had totally forgot about this discussion. A see a few small issues right now:

1. If it's not enabled by default - few people will likely use it
2. It's still won't help people running RuboCop on CIs (as often they use one-off VMs)
3. This should be extensively documented in the README

bbatsov commented on lib/rubocop/options.rb:
```diff
@@ -57,6 +57,9 @@ def validate_compatibility
         fail ArgumentError, 'Lint/UnneededDisable can not be used with --only.'
       end
+      if @options.key?(:cache) && @options.key?(:auto_correct)
+        fail ArgumentError, '--cache can not be used with --auto-correct.'
```
Why so? Can't we just invalidate the things we auto-correct?

jonas054 replied:
Yes, probably. Either way, it makes no sense to have this check as caching could also be turned on in configuration.

jonas054 commented:
Good feedback. I'll make it on by default, but then I also need to add a `--no-cache` option. And update README. It's unfortunate that CI runs typically won't benefit. That means we also need come up with some good old-fashioned optimizations. But that's another topic.

jonas054 commented:
Working on a solution that addresses the issues raised here. I have also identified a couple of more problems.
* Running without `-D`, saving to the cache, and then with `-D` reading from the cache, you get no cop names in the output. I'm sure the problem is more general, i.e., options are not handled correctly in cache invalidation.
* The configuration parameter `AllCops: MaxFilesInCache` is global, so if one inspected project sets it really low, it causes cached results to be erased for other projects. One solution could the to store the cache under the current directory, in `.rubocop_cache/` instead of under `/tmp/rubocop_cache/`.

jonas054 commented:
Pushed some fixes.

bbatsov commented on README.md:
```diff
@@ -811,6 +813,23 @@ RuboCop::RakeTask.new(:rubocop) do |task|
+
+Large projects containing hundreds or even thousands of files can take
+a really long time to inspect, but RuboCop has functionality to
+mitigate this problem. A result caching mechanism is activated if the
```
I think the term "result" is kind of confusing/unclear. I'd suggest elaborating on this.

bbatsov commented on README.md:
```diff
@@ -811,6 +813,23 @@ RuboCop::RakeTask.new(:rubocop) do |task|
+default. The command line option `--cache false` can be used to turn
+off result caching, thus overriding the configuration parameter. If
+`AllCops: UseCache` is set to `false` in the local `.rubocop.yml`,
+then it's `--cache true` that overrides the setting.
```
There should also be explanations about the things which would invalidate the cache.

bbatsov commented on README.md:
```diff
@@ -811,6 +813,23 @@ RuboCop::RakeTask.new(:rubocop) do |task|
+`AllCops: UseCache` is set to `false` in the local `.rubocop.yml`,
+then it's `--cache true` that overrides the setting.
+
+The cache is stored in `.rubocop_cache/` under the current directory.
```
That'd be a bad idea with linters, as they would litter the entire project tree with such folders.

kbrock replied:
Is this cache folder created at the root, or in every folder?

I thought it would be in the root directory, although, that may be tricky to find.

kbrock commented on README.md:
```diff
@@ -811,6 +813,23 @@ RuboCop::RakeTask.new(:rubocop) do |task|
+The cache is stored in `.rubocop_cache/` under the current directory.
+There are subdirectories for each user, so as to minimize the risk of
+collisions and file curruption from simultaneous `rubocop`
+invocations. Git users should add `.rubocop_cache` to their
```
if there are multiple simultaneous `rubocop` runs, it will probably be run by the same user.

- A cruise control system will run as 1 user, and a different directory per branch. This affects a common temp directory but not a local `.rubocop_cache` directory.
- Local developer will run as 1 user. Guard being the use case that may run the same files in same directories, but that seems unlikely.
- Travis will run on different pristine directories, (so caching won't help)

To best handle a race condition, a new cache file could be written and renamed to the target name. last one in wins is better than a corrupt file. But not sure if this really needs to be handled.

bbatsov replied:
I don't think we've factored in simultaneous RuboCop runs so far. I really don't want to have the cache stored in the actually directories where rubocop is invoked. People should also keep in mind that while normally people invoke it on a single directory or file there's nothing stopping you from invoking it on a ton of unrelated folders. Guess a simple solution would be to simply monitor the target directory structure under `/temp` or something like this and to create some locks while updating the cache.

jonas054 replied:
@bbatsov Yes, I will try to implement something along those lines. I like @kbrock's idea about renaming files as a way to make the operation atomic. I think it could work, and I'm going to try it.

wli replied:
@jonas054 any chance you could make the cache directory configurable in an option? CircleCI allows us to cache specific directories between runs (maybe Travis has something similar?). We do this now for rails asset caching - speeds up an asset compile from 2+ minutes to 10 seconds, and CircleCI stores the cache per-branch.

I don't think CircleCI will let us cache anything out of /tmp though, since I think it only reloads things under the project/user directory. I suppose we could always work around this by copying the cache directory, but that's less than ideal.

For reference, this is what our CircleCI config looks like to achieve this:
```yml
dependencies:
  pre:
    - mkdir -p public/assets
    - mkdir -p tmp/cache/assets
  post:
    - bundle exec rake assets:precompile:
        environment:
          RAILS_ENV: test
  cache_directories:
    - public/assets/
    - tmp/cache/assets
```

jonas054 replied:
@wli Sounds like a good idea to me. I'll try to include a configuration parameter for cache root.

jonas054 commented:
Pushed some new updates. Please check again.

bbatsov commented:
Looks fine to me.

bbatsov commented:
Btw, one idea for optimization for dealing with config changes or CLI option changes would be to also cache the ASTs, so we can reuse them directly in such scenarios.

jonas054 commented:
I haven't tested caching the ASTs, so I can't say for sure, but I suspect that it won't do much for performance.

So if we push that aside, maybe we're ready to merge?

bbatsov commented:
Fair enough. Let's merge this!

wli commented:
:+1: Can't wait to use this!
